mongodb-plugins/src/main/java/co/cask/plugin/etl/batch/source/MongoDBBatchSource.java

```diff
@@ -0,0 +1,187 @@
+  "into a StructuredRecord with the help of the specified Schema. " +
+  "Optionally, the user can specify input query, input fields, and splitter class.")
+public class MongoDBBatchSource extends BatchSource<Object, BSONObject, StructuredRecord> {
+
+  private static final Gson GSON = new Gson();
+  private final MongoDBConfig config;
+
+  public MongoDBBatchSource(MongoDBConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws Exception {
+    Job job = context.getHadoopJob();
+    Configuration conf = job.getConfiguration();
+    MongoConfigUtil.setInputFormat(conf, MongoInputFormat.class);
+    MongoConfigUtil.setInputURI(conf, config.connectionString);
+    if (!Strings.isNullOrEmpty(config.inputQuery)) {
+      MongoConfigUtil.setQuery(conf, config.inputQuery);
+    }
+    if (!Strings.isNullOrEmpty(config.authConnectionString)) {
+      MongoConfigUtil.setAuthURI(conf, config.authConnectionString);
+    }
+    if (!Strings.isNullOrEmpty(config.inputFields)) {
+      MongoConfigUtil.setFields(conf, config.inputFields);
+    }
+    if (!Strings.isNullOrEmpty(config.splitterClass)) {
+      String className = String.format("%s.%s", StandaloneMongoSplitter.class.getPackage().getName(),
+                                       config.splitterClass);
+      Class<? extends MongoSplitter> klass = getClass().getClassLoader().loadClass(
+        className).asSubclass(MongoSplitter.class);
+      MongoConfigUtil.setSplitterClass(conf, klass);
+    }
+    Schema.parseJson(config.schema);
+    job.setInputFormatClass(MongoConfigUtil.getInputFormat(conf));
+  }
+
+  @Override
+  public void transform(KeyValue<Object, BSONObject> input, Emitter<StructuredRecord> emitter) throws Exception {
+    BSONObject bsonObject = input.getValue();
+    Schema schema = Schema.parseJson(config.schema);
+    StructuredRecord.Builder builder = StructuredRecord.builder(schema);
+    for (Schema.Field field : schema.getFields()) {
+      builder.set(field.getName(), convertValue(bsonObject.get(field.getName()).toString(), field));
+    }
+    emitter.emit(builder.build());
+  }
+
+  private Object convertValue(String input, Schema.Field field) throws IOException {
+    switch (field.getSchema().getType()) {
+      case NULL:
+        return null;
+      case BOOLEAN:
+        return Boolean.valueOf(input);
+      case INT:
+        return Integer.valueOf(input);
+      case LONG:
+        return Long.valueOf(input);
+      case FLOAT:
+        return Float.valueOf(input);
+      case DOUBLE:
+        return Double.valueOf(input);
+      case BYTES:
+        return input.getBytes();
+      case STRING:
+      case ENUM:
+        return input;
+      case ARRAY:
+        return GSON.fromJson(input, List.class);
+      case MAP:
+        return GSON.fromJson(input, Map.class);
+      case UNION:
+        if (field.getSchema().isNullableSimple()) {
```

Comment:
this means nullable arrays and maps are not allowed. Also, if input is null, this will end up in a NPE.